Adding EE/IG/labels/forks/timeout/job_slice_count to schedules

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
TODO:
- [x] AWX Collection Changes
- [x] New unit tests
- [x] Test WFJT Nodes


<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature
 
##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 
##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Adding EE/IG/labels/forks/timeout/job_slice_count to schedules

john-westcott-iv committed:
Adding Instance Groups to schedules

john-westcott-iv committed:
Adding clear field on schedule many2many fields

john-westcott-iv committed:
Allowing for unsaved objects in validate method

john-westcott-iv committed:
Fixing migration file

john-westcott-iv committed:
Fixing stack if a new schedule is added

john-westcott-iv committed:
Modifying schedules to work with related fields

john-westcott-iv committed:
Fixing schedules module and related issues with schedules

john-westcott-iv committed:
Working on WorkflowJob[Template]Node

john-westcott-iv committed:
Updating awx.awx.workflow_job_template_node

mabashian committed:
Cleans up UI linting errors

mabashian committed:
Fix broken UI unit tests

mabashian committed:
Adds missing LabelsMixin import

john-westcott-iv committed:
Fixing tests

john-westcott-iv committed:
Adding unit/functional tests

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -2985,6 +3049,38 @@ class WorkflowJobTemplateNodeCredentialsList(LaunchConfigCredentialsBase):
+            return Response(
+                dict(msg=_('Maximum number of labels for {} reached.'.format(self.parent_model._meta.verbose_name_raw))), status=status.HTTP_400_BAD_REQUEST
+            )
+        return super(WorkflowJobTemplateNodeLabelsList, self).post(request, *args, **kwargs)
```
since this has already repeated 3 times (that you added), this would be a good candidate to consolidate into a shared method.

john-westcott-iv replied:
Done.

john-westcott-iv committed:
Making common class for LabelList

john-westcott-iv committed:
Fixing related field name

john-westcott-iv committed:
Fixing get_effective_slice_ct to look for corerct field and also override _eager_field

mabashian committed:
Extends LabelSelect to have a custom chip render.  This allows us to disable labels that cannot be removed on job launch

AlanCoding committed:
Add WorkflowJob.instance_groups and distinguish from char_prompts

This removes a loop that ran on import
  the loop was giving the wrong behavior
  and it initialized too many fields as char_prompts fields

With this, we will now enumerate the char_prompts type fields manually

john-westcott-iv committed:
Merge pull request #49 from AlanCoding/wj_igs

Add WorkflowJob.instance_groups and distinguish from char_prompts

john-westcott-iv committed:
Making labels additive and not adding a many item to config if already in parent

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -209,6 +209,26 @@ def api_exception_handler(exc, context):
     return exception_handler(exc, context)
 
 
+class LabelList(DeleteLastUnattachLabelMixin, SubListCreateAttachDetachAPIView):
```
class name is a little confusing, would be better to say `LabelSubList` or something /nit

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -209,6 +209,26 @@ def api_exception_handler(exc, context):
+    def post(self, request, *args, **kwargs):
+        # If a label already exists in the database, attach it instead of erroring out
+        # that it already exists
+        if not getattr(self, 'label_filter', None):
```
If it's a reverse ForeignKey then `parent_key` is the back-reference. It seems that we don't have something similar for many-to-many relationship sublists.

But it seems that we could get it in an algorithmic way.

```
In [6]: Schedule._meta.get_field('labels').remote_field.name
Out[6]: 'schedule_labels'

In [7]: JobTemplate._meta.get_field('labels').remote_field.name
Out[7]: 'unifiedjobtemplate_labels'

In [8]: WorkflowJobNode._meta.get_field('labels').remote_field.name
Out[8]: 'workflowjobnode_labels'
```

It's not that I'd prefer this generally, but here, it would avoid the need for the `return Response(...)` part, which is not good practice and also fairly unreachable by testing.

AlanCoding commented on awx/main/access.py:
```diff
@@ -1923,19 +1923,91 @@ def can_change(self, obj, data):
+                    "User {} not allowed access to instance_group {} for {} {} ({})".format(self.user.username, sub_obj.name, obj.__class__, obj_name, obj.id)
+                )
+                return False
+            return True
```
I get the need for some divergences from the parent class here:

https://github.com/ansible/awx/blob/9203062cd64b929cbc6d692066aec7d2e58ab63f/awx/main/access.py#L284-L288

But I'd like to sketch out the goals here from the first principles. In this change, you added `instance_groups` to WFJT nodes, not just schedules. This is evidenced by the introduction of the view `WorkflowJobTemplateNodeInstanceGroupsList`, likewise with schedules.

But WFJT node access class makes no reference to the access method here. So why bother? Or is it broken?

There's also a high degree of repetition here. If the node identifier is a big problem, then fix it in `BaseAccess.can_attach`, and it's really hard for me to see the need for the subclass to handle much... other than not checking the primary object can_change (because it's a supplemental check), but this makes it hard to justify keeping this method at all, since the `BaseAccess` implementation will probably be fine as it is manifested for `ScheduleAccess.can_attach` and the like.

AlanCoding reviewed (approved):
LGTM into the feature branch, recognizing that the RBAC logic still needs some work and testing.
